src/Storages/PostgreSQL/PostgreSQLReplicaConsumer.cpp

```diff
@@ -0,0 +1,682 @@
+                        "For table engine full replica identity is also supported. Table will be skipped.");
+                markTableAsSkipped(relation_id, relation_name);
+                return;
+            }
+
+            Int16 num_columns = readInt16(replication_message, pos, size);
+
+            Int32 data_type_id;
+            Int32 type_modifier; /// For example, n in varchar(n)
+
+            bool new_relation_definition = false;
+            if (schema_data.find(relation_id) == schema_data.end())
+            {
+                relation_id_to_name[relation_id] = relation_name;
+                schema_data.emplace(relation_id, SchemaData(num_columns));
+                new_relation_definition = true;
+            }
+
+            auto & current_schema_data = schema_data.find(relation_id)->second;
+
+            if (current_schema_data.number_of_columns != num_columns)
+            {
+                markTableAsSkipped(relation_id, relation_name);
+                return;
+            }
+
+            for (uint16_t i = 0; i < num_columns; ++i)
+            {
+                String column_name;
+                readInt8(replication_message, pos, size); /// Marks column as part of replica identity index
+                readString(replication_message, pos, size, column_name);
+
+                data_type_id = readInt32(replication_message, pos, size);
+                type_modifier = readInt32(replication_message, pos, size);
+
+                if (new_relation_definition)
+                {
+                    current_schema_data.column_identifiers.emplace_back(std::make_tuple(data_type_id, type_modifier));
+                }
+                else
+                {
+                    if (current_schema_data.column_identifiers[i].first != data_type_id
+                            || current_schema_data.column_identifiers[i].second != type_modifier)
+                    {
+                        markTableAsSkipped(relation_id, relation_name);
+                        return;
+                    }
+                }
+            }
+
+            if (storages.find(relation_name) == storages.end())
+            {
+                throw Exception(ErrorCodes::LOGICAL_ERROR,
+                        "Storage for table {} does not exist, but is included in replication stream", relation_name);
+            }
+
+            [[maybe_unused]] auto buffer_iter = buffers.find(relation_name);
+            assert(buffer_iter != buffers.end());
+
+            tables_to_sync.insert(relation_name);
+
+            break;
+        }
+        case 'O': // Origin
+            break;
+        case 'Y': // Type
+            break;
+        case 'T': // Truncate
+            break;
+        default:
+            throw Exception(ErrorCodes::LOGICAL_ERROR,
+                    "Unexpected byte1 value {} while parsing replication message", type);
+    }
+}
+
+
+void PostgreSQLReplicaConsumer::syncTables(std::shared_ptr<pqxx::nontransaction> tx)
```

Comment:
It's called only after we found `R` message in replication log, right? Maybe we should also flush buffer if it contains more than `some_setting_name` rows or timeout of `some_other_settings_name` ms elapsed

Reply:
It is already so, It is flushed after reading `max_block_size` messages. (In fact can be more in case one transaction contained more than `max_block_size` rows changed, in this case it will be all this transaction and then flush).

Reply:
Hm, but we insert something into `tables_to_sync` only [here](https://github.com/ClickHouse/ClickHouse/blob/5647c01be49003c15b20de4eb7f275aa012e4493/src/Storages/PostgreSQL/PostgreSQLReplicaConsumer.cpp#L634) and [clear](https://github.com/ClickHouse/ClickHouse/blob/5647c01be49003c15b20de4eb7f275aa012e4493/src/Storages/PostgreSQL/PostgreSQLReplicaConsumer.cpp#L515) it at the end of `syncTables(...)`, so `tables_to_sync` can be [not empty](https://github.com/ClickHouse/ClickHouse/blob/5647c01be49003c15b20de4eb7f275aa012e4493/src/Storages/PostgreSQL/PostgreSQLReplicaConsumer.cpp#L634) only after processing of `R` message. Am I missing something?

Reply:
Yes, table names are added into `tables_to_sync` only after there was `'R'` message for them. Because there is always this type of message for each table included in current messages stream. (It is received for the first time before the first message (`update`, `insert`, ..) for according table and can be received in current stream more times - only in case table structure was changed).

Reply:
Oh, I understood what you meant.

Reply:
So you suggest when checking each table in `tables_to_sync` (after each replication messages stream) I should also check its buffer size and timeout and sync each separate table only if needed?